app/scripts/ui.js

```diff
@@ -36,8 +40,36 @@ const WORKER_KEEP_ALIVE_MESSAGE = 'WORKER_KEEP_ALIVE_MESSAGE';
  * Time has been kept to 1000ms but can be reduced for even faster re-activation of service worker
  */
 if (isManifestV3) {
-  setInterval(() => {
+  const handle = setInterval(() => {
+    const messageId = new Date().getTime() + Math.random();
+    ACK_KEEP_ALIVE_RECEIVED[messageId] = false;
     browser.runtime.sendMessage({ name: WORKER_KEEP_ALIVE_MESSAGE });
+
+    /**
+     * set timeout to clear the ACK_KEEP_ALIVE_RECEIVED or
+     * show error if response is not received after ACK_KEEP_ALIVE_WAIT_TIME seconds
+     */
+    const timeoutHandle = setTimeout(() => {
+      if (ACK_KEEP_ALIVE_RECEIVED[messageId]) {
+        delete ACK_KEEP_ALIVE_RECEIVED[messageId];
+      } else {
+        clearInterval(handle);
+        displayCriticalError(
+          'somethingIsWrong',
+          new Error("Something's gone wrong. Try reloading the page."),
+        );
+      }
+    }, ACK_KEEP_ALIVE_WAIT_TIME);
+
+    // add listener to receive ACK_KEEP_ALIVE_MESSAGE
+    // eslint-disable-next-line no-undef
+    const channel = new window.BroadcastChannel('sw-messages');
+    channel.addEventListener('message', (event) => {
+      if (event.data.name === ACK_KEEP_ALIVE_MESSAGE) {
+        clearTimeout(timeoutHandle);
+        delete ACK_KEEP_ALIVE_RECEIVED[messageId];
+      }
+    });
```

Comment:
Do we need a new channel here. We already have message listeners in this file.

Reply:
The channel is to respond to the UI. It's only method that worked I tried the below

1. I used the sender argument of addListener callback, message is not going back to the client. One thing I noticed though (not sure if this is the reason). If I get all the clients and print their IDs, there's no ID for the sender argument. Basically I can't match the sender to any of the connected clients.
2. I also used the sendReponse method of the addListener callback, same issue with sender argument. The message is not coming back to the UI
3. The only method that worked is the BroadcastChannel (a MessageChannel would work too, but it's more complex)

Reply:
perhaps we could use the the portstreams already established between the background and the ui script?

Reply:
At this point, there are no port streams initialised yet. The major issue is I can't find a way to match the client-id else, there's a client.postMessage that could be called within the event handler. But like I said above
```
If I get all the clients and print their IDs, there's no ID for the sender argument. Basically I can't match the sender to any of the connected clients.
```

Reply:
@danjm @jpuri probably we can re-visit this after [this](https://github.com/MetaMask/metamask-extension/pull/16075) is merged.

In the above linked PR, I see that the keep alive message is only sent after the phishing stream have been initialized

```
 if (isManifestV3) {
     phishingPageStream.on('data', ({ data: { method } }) => {
       if (!IGNORE_INIT_METHODS_FOR_KEEP_ALIVE.includes(method)) {
         runWorkerKeepAliveInterval();
       }
     });
   }
```